File: 面试题集锦/全排序/permute.go
package main

import (
	"fmt"
)

func main() {
	str := "abc"
	length := len(str)
	result := []string{}
	permute("", str, length, &result)
	fmt.Println(result)
}

func permute(combin string, rest string, max int, result *[]string) {
	if len(combin) == max {
		*result = append(*result, combin)
		return
	}
	for i, _ := range(rest) {
		permute(combin+string(rest[i]), rest[:i]+rest[i+1:], max, result)
	}
}

File: 面试题集锦/全排序/permute.py
#coding=utf-8

def permute(string):
    res = list()
    length = len(string)
    def combin(coms, rest):
        if len(coms) == length:
            res.append(coms)
            return
        for i in range(len(rest)):
            combin(coms+rest[i], rest[:i]+rest[i+1:])
    combin("", string)
    return res

if __name__ == "__main__":
    string = input("输入字符：")
    print(permute(string))
